subprojects/workers/src/integTest/groovy/org/gradle/process/internal/WorkerDaemonServiceErrorHandlingIntegrationTest.groovy

```diff
@@ -0,0 +1,327 @@
+
+        and:
+        failureHasCause("A failure occurred while executing org.gradle.test.TestRunnable")
+
+        and:
+        failureHasCause("Failed to run Gradle Worker Daemon")
+    }
+
+    def "produces a sensible error when a parameter can't be serialized"() {
+        withRunnableClassInBuildSrc()
+        withUnserializableParameterInBuildSrc()
+
+        buildFile << """
+            $alternateRunnable
+
+            task runAgainInDaemon(type: DaemonTask) {
+                runnableClass = AlternateRunnable.class
+            }
+            
+            task runInDaemon(type: DaemonTask) {
+                foo = new FooWithUnserializableBar()
+                finalizedBy runAgainInDaemon
+            }
+        """
+
+        when:
+        fails("runInDaemon")
+
+        then:
+        failureHasCause("A failure occurred while executing org.gradle.test.TestRunnable")
+        failureCauseContains("Could not write message")
+        errorOutput.contains("Caused by: java.io.NotSerializableException: org.gradle.error.Bar")
+
+        and:
+        executedAndNotSkipped(":runAgainInDaemon")
+        assertRunnableExecuted("runAgainInDaemon")
+    }
+
+    def "produces a sensible error when a parameter can't be de-serialized in the worker"() {
+        def parameterJar = file("parameter.jar")
+        withRunnableClassInBuildSrc()
+        withUnserializableParameterMemberInExternalJar(parameterJar)
+
+        buildFile << """  
+            $alternateRunnable
+
+            task runAgainInDaemon(type: DaemonTask) {
+                runnableClass = AlternateRunnable.class
+            }
+
+            task runInDaemon(type: DaemonTask) {
+                additionalClasspath = files('${parameterJar.name}')
+                foo = new FooWithUnserializableBar()
+                finalizedBy runAgainInDaemon
+            }
+        """
+
+        when:
+        fails("runInDaemon")
+
+        then:
+        failureHasCause("A failure occurred while executing org.gradle.test.TestRunnable")
+        failureCauseContains("Could not read message")
+        errorOutput.contains("Caused by: java.lang.ClassNotFoundException: org.gradle.error.Bar")
+
+        and:
+        executedAndNotSkipped(":runAgainInDaemon")
+        assertRunnableExecuted("runAgainInDaemon")
+    }
+
+    def "produces a sensible error even if the action failure cannot be fully serialized"() {
+        withRunnableClassInBuildSrc()
+
+        buildFile << """
+            $alternateRunnable
+
+            task runAgainInDaemon(type: DaemonTask) {
+                runnableClass = AlternateRunnable.class
+            }
+
+            $runnableThatThrowsUnserializableMemberException
+
+            task runInDaemon(type: DaemonTask) {
+                runnableClass = RunnableThatFails.class
+                finalizedBy runAgainInDaemon
+            }
+        """
+
+        when:
+        fails("runInDaemon")
+
+        then:
```

Comment:
We should also assert that the thrown unserializable exception is logged.
Something like `errorOutput.contains 'Unserializable exception from runnable'`